### PR TITLE
remove trainer hidden state | sanity refactor [2 / n]

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Follow up to #7437 with more writable state removed from trainer:
- Trainer.hiddens
- Trainer.split_idx
- multiple_trainloader_mode does not belong in TrainLoop

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/core/lightning.py [hunk 1]
```diff
@@ -1619,7 +1619,7 @@ def get_progress_bar_dict(self):
         module_tbptt_enabled = self.truncated_bptt_steps > 0
         trainer_tbptt_enabled = self.trainer.truncated_bptt_steps is not None and self.trainer.truncated_bptt_steps > 0
         if module_tbptt_enabled or trainer_tbptt_enabled:
-            tqdm_dict["split_idx"] = self.trainer.split_idx
+            tqdm_dict["split_idx"] = self.trainer.train_loop.split_idx
```
Comment (ananthsub) [inline pytorch_lightning/core/lightning.py:1622]: is `get_progress_bar_dict` still needed on the lightning module? wouldn't things come from `self.log` ?

File: pytorch_lightning/trainer/training_loop.py [hunk 2]
```diff
@@ -53,17 +52,17 @@ def __init__(
...
         self._optimizer_freq_cumsum = None
+        self._hiddens = None
 
         self.global_step = 0
         self.current_epoch = 0
         self.trainer.should_stop = False
 
         self.total_batch_idx = 0
         self.batch_idx = 0
+        self.split_idx = None
```
Comment (ananthsub) [inline pytorch_lightning/trainer/training_loop.py:65]: - is this meant to be writable from the outside? should these be made available as properties instead? 
- while going through this, could you add a comment for what split_idx means for contributors who go through this code later?

File: pytorch_lightning/trainer/training_loop.py [hunk 3]
```diff
@@ -53,17 +52,17 @@ def __init__(
         self.running_loss = TensorRunningAccum(window_length=20)
         self._curr_step_result = None
         self._cur_grad_norm_dict = None
-        self._multiple_trainloader_mode = multiple_trainloader_mode
         self._skip_backward = False
-        self.trainer._multiple_trainloader_mode = multiple_trainloader_mode
         self._optimizer_freq_cumsum = None
+        self._hiddens = None
```
Comment (ananthsub) [inline pytorch_lightning/trainer/training_loop.py:57]: typehints will also be helpful

File: pytorch_lightning/core/lightning.py (hunk 1 above)
Comment (awaelchli) [reply pytorch_lightning/core/lightning.py:1622]: it's for users to override and customize the default elements in the progress bar. Like how the version number is displayed or apparently the split index  here.

Comment (awaelchli) [reply pytorch_lightning/core/lightning.py:1622]: It's not something one could customize through self.log directly I would say.

Comment (ananthsub) [reply pytorch_lightning/core/lightning.py:1622]: Is it customization we would want to push to a custom progress bar callback instead of being part of the core module?

Comment (carmocca) [reply pytorch_lightning/core/lightning.py:1622]: > custom progress bar callback instead of being part of the core module?
IMO yes for maximum separation of concerns

But people might complain about having to define a custom progress bar just for this

File: pytorch_lightning/trainer/training_loop.py [hunk 4]
```diff
@@ -686,6 +684,7 @@ def run_training_batch(self, batch, batch_idx, dataloader_idx):
         splits = self._tbptt_split_batch(batch)
 
         for split_idx, split_batch in enumerate(splits):
+            self.split_idx = split_idx
```
Comment (carmocca) [inline pytorch_lightning/trainer/training_loop.py:687]: Do you think it's better to have this here instead of in `run_train_split_start`?

Comment (awaelchli) [reply pytorch_lightning/trainer/training_loop.py:687]: Yes, two reasons 1) because this is where the loop starts so if we are tracking split_idx as instance variable, this is the place where we update it 2) we do the same for batch_idx one loop level above.

File: pytorch_lightning/trainer/training_loop.py (hunk 3 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/training_loop.py:57]: is it ok if we add typehints after major refactors?

File: pytorch_lightning/trainer/training_loop.py (hunk 2 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/training_loop.py:65]: no but for now I follow the above pattern and the major goal is to define this state strictly on the loop, not the trainer anymore.

Comment (awaelchli) [reply pytorch_lightning/trainer/training_loop.py:65]: > while going through this, could you add a comment for what split_idx means for contributors who go through this code later?

yes

File: pytorch_lightning/core/lightning.py (hunk 1 above)
Comment (awaelchli) [reply pytorch_lightning/core/lightning.py:1622]: one could argue maybe the split_idx is not very useful to display in the progbar but still I would keep the hook

File: pytorch_lightning/trainer/training_loop.py (hunk 2 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/training_loop.py:68]: okay added comments

**Review by carmocca — APPROVED** [review]

**Review by Borda — APPROVED** [review]

**Review by ananthsub — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
LGTM !
